### DISCUSSION TITLE
利用 Hazel 将 textbundle 修改为 Obsidian 可以预览的文件

### DISCUSSION THREAD

**Discussion by Kenshin** [Kenshin/simpread#3241] [服务]
> 如果你是 Mac + Obsidian 用户的话，可以试试此方式，利用 Hazel 的自动化方便的将简悦导出的 `.textbundle` 修改为 Obsidian 可以查看的结构。

https://user-images.githubusercontent.com/81074/147432736-ff10c2de-ed2b-422a-a55d-31038f4325db.mp4

## 场景需要

Obsidain 无法直接读取  `.textbundle` 格式，而离线 Markdown 因过大的 size 会让 Obsidian 出现假（卡）死问题。

所以为了解决这个问题，简悦目前有以下几种用法：

- [巧用 Textbundle 格式让 Obsidian 更好的支持「包含图片子文件夹的 Markdown」](https://github.com/Kenshin/simpread/discussions/2360)

- [[uTools\]用 Python 脚本转换 Markdown 内嵌的 Base64 为图片（主要在 Logseq 中使用）](https://github.com/Kenshin/simpread/discussions/3237)

- [[Quicker 动作\]批量下载 Markdown 中的外链图片并替换成本地离线图片链接（主要应用场景：通过简悦导出 Markdown 到 Obsidian 中）](https://github.com/Kenshin/simpread/discussions/2241)

- [[Quicker 动作\]用 Quicker 转换 Obsidian 中的离线 Markdown 文档中的图片](https://github.com/Kenshin/simpread/discussions/3132)

## 自动化方案

前两天  [简悦社区](https://t.me/simpreadgroup) 用户 [FtgsgG](https://www.zhihu.com/people/xu-hao-84-8) 带来了一个 [教程](https://github.com/Kenshin/simpread/discussions/3235)，利用 Hazel 的自动化方案实现。

在跟 FtgsgG 沟通时，想到能否利用 Hazel 的自动化功能将手动修改  `.textbundle` 的方案改为自动化方式。

所以，就有了他的第二篇文章：

- [简悦导出为 textbundle 格式后的文件名修正](https://zhuanlan.zhihu.com/p/449867382)

此方式利用了 Hazel 可以运行脚本的功能：

![image](https://user-images.githubusercontent.com/81074/147433637-735d1e68-891a-44b8-bbe6-f0590aba116a.png)

### 使用流程

1. 监控 output 文件夹
2. 当存在 `.textbundle` 文件夹时，自动完成：删除 `.textbundle`、删除 `info.json`、改名 的流程。

### 核心代码

```
string=$1

# 分割成 xxx.textbundle
temp=${string##*\/}

# 提取文章名 xxx
article_name=${temp%%.*}

# Markdown 文件之前的名字
name_before=$string"/text.markdown"

# 重命名之后的 Markdown 名字
name_after=$string"/$article_name.md"

# 重命名 Markdown
mv $name_before $name_after

# 重命名后的文件夹名
file_name=${string%%.*}

# 重命名文件夹
mv $string $file_name

# 得到 info.json的地址
json=$file_name"/info.json"

# 删除 info.json
rm -f $json
```

## 进一步的方案

FtgsgG 的方案是将 Obsidian 库文件夹设置为 `output` 文件夹的前提下，一般来说 Obsidian 库文件夹与 `output`  大多都不在一个地方。

所以，可以将上面的方案修改为下面的方式。

![image](https://user-images.githubusercontent.com/81074/147434201-f45c6a9e-5560-4b46-b9fb-9d26ced4cef4.png)

> 简单的解释下上图标号对应的内容：

1.  仅监控 `.textbundle` 文件

2. 因为`.textbundle` 内含了本地图片，所以当发生保存时有个保存过程，所以需要 **这个文件产生一分钟之后** 再执行自动化。

3. 当符合条件时，触发自动化脚本（也就是上面的那段代码）

4. 将这个文件转移到 Obsidian 库文件夹中

5. 为这个自动化方案起个名字

6. 监控 `ouput` 文件夹

## Hazel

Hazel 是一个 Mac 下效率工具，主要功能是监控文件夹的变化，从而进行一系列自动化，类似本地 IFTTT 的概念。

Hazel 能做的事情还很多，这里只是抛砖引玉了其中一种。

## Windows 用户

Windows 用户可使用 DropIt ，细节请看  [利用 DropIt 的文件夹监控功能，助力简悦标注自动导出到 Logseq（Windows 方案）](https://github.com/Kenshin/simpread/discussions/3469)](https://github.com/Kenshin/simpread/discussions/3469)

## 关联

- [简悦 + Logseq + Github Page 无代码全自动化知识管理发布方案](https://github.com/Kenshin/simpread/discussions/3426)

- [利用 Keyboard Maestro 的文件夹监控功能，将简悦标记导入到 Logseq](https://github.com/Kenshin/simpread/discussions/3581)

- [利用 DropIt 的文件夹监控功能，助力简悦标注自动导出到 Logseq](https://github.com/Kenshin/simpread/discussions/3469)

**Comment by ernestyao** [comment]
如果只是放到inbox，后续整理的时候仍然需要调整图片链接（如果同名，容易引起obsidian引用错误）。要么就变成一篇网帖一个目录，和obisidan一般管理习惯不同。所以我稍微改造了一下脚本，实现把附件文件夹改名为和文件同名、统一放到一个地方（脚本里是attachments），同时把markdown正文里的图片链接地址也替换掉。

```
string=$1

# 分割成 xxx.textbundle
temp=${string##*\/}

# 提取文章名 xxx
article_name=${temp%%.*}
folder_name=`echo $article_name | sed 's/ //g'`

# Markdown 文件之前的名字
name_before=$string"/text.markdown"

# 重命名之后的 Markdown 名字
name_after="$article_name.md"

# 重命名 Markdown
mv "$name_before" ./"$name_after"

# 重命名文件夹
mv "$string"/assets attachments/$folder_name

# 得到 info.json的地址
json=$string"/info.json"

# 删除 info.json
rm -f $json

# 重命名markdown中附件路径
sed -i '' -e "s|assets/|attachments/$folder_name/|g" "$name_after"
```

我的教程 https://www.readern.com/import-textbundle-to-obsidian.html

**Reply by Kenshin** [reply]
Cool~

明天放到 https://t.me/simpread 以及简悦各种渠道推广下。

**Reply by scramjetFoam** [reply]
我遇到一个问题，通过此方法可以完美分离textbundle文件为md文件和attachments子图片。但是因为我设置了标注的自动化动作，后续标注时，更新的md文件会替换掉原textbundle转换的md文件。不知道该怎么解决？我的想法是给textbundle转换md文件时加个后缀以区分，但是不知道如何修改脚本。

**Reply by Kenshin** [reply]
> 我遇到一个问题，通过此方法可以完美分离textbundle文件为md文件和attachments子图片。但是因为我设置了标注的自动化动作，后续标注时，更新的md文件会替换掉原textbundle转换的md文件。不知道该怎么解决？我的想法是给textbundle转换md文件时加个后缀以区分，但是不知道如何修改脚本。

标注的自动化可以将文件导出为 `xxx@annote.md` 的文件这样就不会存在这个问题了，如何配置看 [知识库](https://kb.simpread.pro/#/page/%E5%BB%BA%E7%AB%8B%E7%9F%A5%E8%AF%86%E5%BA%93)

**Reply by scramjetFoam** [reply]
好的，谢谢解答。好像是因为我稍后读自动化同时设置了导出markdown和textbundle，然后hazel的自动把textbundle转为了markdown。现在我是这样配置的，应该没问题了。
<img width="876" alt="CleanShot 2022-04-25 at 14 14 24@2x" src="https://user-images.githubusercontent.com/80505084/165030811-65f075d0-e766-4d07-9e58-7c626a4ee2eb.png">

**Reply by Kenshin** [reply]
@ShangWangNUST 

Cool，搞定就好~

**Comment by Kenshin** [comment]
嗨，各位。

简悦官方针对新用户或不喜欢配置的用户，发布了一个新功能：**Obsidan 配置库**。

细节请看下面，几分钟就能配置成完成。

***

![img](https://imgs.zhubai.love/c626678ff6b64dd1ac9dfdae6cf6d188.png)

> 针对新用户的一步到位的 Obsdian 配置库。

## 🗂 配置库

[简悦 · 配置库](https://www.yuque.com/kenshin/simpread/ds8zk0) 是简悦官方推出的一套针对新用户的极简配置方案，方便新用户用最快的方式使用简悦的各种高级服务，配置库内置了常用的双链笔记用法，如：Notion、Obsidian、Logseq、Roam Research，同时包含了简悦在阅读模式上的一些常规插件：Live Editor、题图、Safari 阅读模式等。

## 💡 特点

用户仅需要下载对应的配置库，修改里面的 UID 为自己的高级账户 UID，即可实现简悦的各种高级操作。

## 💁‍♂️ 受众

新用户或者不想使用手动配置的用户。

## 📝 如何使用

1️⃣ 下载

2️⃣ 修改为自己的 UID

3️⃣ 将必要的文件夹复制粘贴到相应位置

✅ Done！

## 📚 Obsidian 配置库

## 📝 特点

1️⃣ 加入稍后读，自动将本地快照和 Markdown 导入到 Obsidian

2️⃣ 加入标注时，自动将标注内容导入本地快照和 Obsidian

3️⃣ 改变稍后读或标注的元数据时，自动将改动的内容保存到本地快照和 Obsidian

4️⃣ 在 Obsidian 中直接内置好了简悦的标注系统，可直接在 Obsidian 实现标注

## 📖 区别

![img](https://imgs.zhubai.love/f4a51d8e8fb4401281b07f82bf0bc7d1.png)

## 📘 极速版

🔗 [Github](https://github.com/Kenshin/simpread-configs/blob/main/obsidian@little/Getting%20Started.md) | [语雀](https://www.yuque.com/kenshin/simpread/xkuecp)

不需要同步助手，无任何配置，支持 Markdown 形式的快照，自动同步标注，适合 Obsidain + 简悦的尝鲜用户。

## 📙 基础版

🔗 [Github](https://github.com/Kenshin/simpread-configs/blob/main/obsidian@localrestapi/Getting%20Started.md) | [语雀](https://www.yuque.com/kenshin/simpread/cg33gh)

需要同步助手，几乎无任何配置，支持本地快照，自动同步标注，适合 Obsidain + 简悦的轻量级用户。

## 📗 高级版

🔗 [Github](https://github.com/Kenshin/simpread-configs/blob/main/obsidian@sync/Getting%20Started.md) | [语雀](https://www.yuque.com/kenshin/simpread/wq35mh)

需要同步助手，几乎无任何配置，支持本地快照，自动同步标注，可在 Obsidian 中标注，适合 Obsidain + 简悦的重度用户。
